espnet2/bin/spk_embed_extract.py

```diff
@@ -0,0 +1,693 @@
+from espnet2.torch_utils.set_all_random_seed import set_all_random_seed
+from espnet2.train.distributed_utils import (
+    DistributedOption,
+    free_port,
+    get_master_port,
+    get_node_rank,
+    get_num_nodes,
+    resolve_distributed_mode,
+)
+from espnet2.train.reporter import Reporter, SubReporter
+from espnet2.utils import config_argparse
+from espnet2.utils.build_dataclass import build_dataclass
+from espnet2.utils.types import (
+    humanfriendly_parse_size_or_none,
+    int_or_none,
+    str2bool,
+    str2triple_str,
+    str_or_none,
+)
+from espnet.utils.cli_utils import get_commandline_args
+
+
+def extract_embed(args):
+    distributed_option = build_dataclass(DistributedOption, args)
+    distributed_option.init_options()
+
+    if not distributed_option.distributed or distributed_option.dist_rank == 0:
+        if not distributed_option.distributed:
+            _rank = ""
+        else:
+            _rank = (
+                f":{distributed_option.dist_rank}/"
+                f"{distributed_option.dist_world_size}"
+            )
+
+        logging.basicConfig(
+            level=args.log_level,
+            format=f"[{os.uname()[1].split('.')[0]}{_rank}]"
+            f" %(asctime)s (%(module)s:%(lineno)d) %(levelname)s: %(message)s",
+        )
+    else:
+        # Suppress logging if RANK != 0
+        logging.basicConfig(
+            level="ERROR",
+            format=f"[{os.uname()[1].split('.')[0]}"
+            f":{distributed_option.dist_rank}/{distributed_option.dist_world_size}]"
+            f" %(asctime)s (%(module)s:%(lineno)d) %(levelname)s: %(message)s",
+        )
+    # Invoking torch.distributed.init_process_group
+    distributed_option.init_torch_distributed()
+
+    logging.basicConfig(
+        level=args.log_level,
+        format="%(asctime)s (%(module)s:%(lineno)d) %(levelname)s: %(message)s",
+    )
+
+    if args.ngpu >= 1:
+        device = "cuda"
+    else:
+        device = "cpu"
+
+    # 1. Set random-seed
+    set_all_random_seed(args.seed)
+
+    # 2. define train args
+    spk_model, spk_train_args = SpeakerTask.build_model_from_file(
+        args.spk_train_config, args.spk_model_file, device
+    )
+
+    # 3. Overwrite args with inference args
+    args = vars(args)
+    args["valid_data_path_and_name_and_type"] = args["data_path_and_name_and_type"]
+    args["valid_shape_file"] = args["shape_file"]
+    args["preprocessor_conf"] = {
+        "target_duration": args["target_duration"],
+        "num_eval": args["num_eval"],
+        "noise_apply_prob": 0.0,
+        "rir_apply_prob": 0.0,
+    }
+
+    merged_args = vars(spk_train_args)
+    merged_args.update(args)
+    args = argparse.Namespace(**merged_args)
+
+    # 4. Build data-iterator
+    distributed_option.distributed = False
+    iterator = SpeakerTask.build_iter_factory(
+        args=args,
+        distributed_option=distributed_option,
+        mode="valid",
+    )
+    distributed_option.distributed = True
```

Comment:
It is better to add some comments here to explain why we set `distributed_option.distributed` manually.